### PR TITLE
fixed PCA align=True problems

### LINKED ISSUE #2510: align=True for analysis.PCA gives mean positions of 0
**Expected behavior**

At some point, pca.PCA computes an average structure, to be subtracted from each frame, to generate a covariance matrix. Also if I set `align=True` it aligns properly.

**Actual behavior**

It gives mean positions of 0.0 instead if `align=True`. This messes with the covariance matrix.

This is because in the code, `self.mean` only gets updated with positions in the `else` condition below. Moving that line out of `else` should fix it.

https://github.com/MDAnalysis/mdanalysis/blob/3e8ea8e0744f3163ddbf7b4f5f66874dd26796a7/package/MDAnalysis/analysis/pca.py#L238-L246

```diff
--- else: 
---     self.mean += self._atoms.positions.ravel() 
+++ self.mean += self._atoms.positions.ravel() 
```

In addition, `_fit_to` requires the mobile atoms  have already been translated to the origin by subtracting `mobile_cog`. PCA neglects to do this in two cases: once in the code above, and once below.

https://github.com/MDAnalysis/mdanalysis/blob/02f35b58f8d2d9f43b6821c7f6110f7137db73f3/package/MDAnalysis/analysis/pca.py#L253-L266

Subtracting the center of geometry from `self._atoms.positions` will translate them to the origin.

```diff
---         mobile_atoms, old_rmsd = _fit_to(self._atoms.positions, 
+++         mobile_atoms, old_rmsd = _fit_to(self._atoms.positions-mobile_cog,
```

**Code to reproduce the behavior**

Show us how to reproduce the failiure. If you can, use trajectory files from the test data.

``` python
>>> import MDAnalysis as mda
>>> from MDAnalysis.tests.datafiles import PSF, DCD
>>> from MDAnalysis.analysis import pca, align

>>> u = mda.Universe(PSF, DCD)
>>> p = pca.PCA(u, select='backbone', align=True).run()
>>> p.mean
array([0., 0., 0., ..., 0., 0., 0.])
>>> p.cov
array([[ 209.12140302,  104.99822377, -126.03359734, ...,  151.80809005,
         239.13639816,  -96.69853061],
       [ 104.99822377,   53.77314286,  -64.13001212, ...,   75.19231419,
         121.77024102,  -49.10982093],
       [-126.03359734,  -64.13001212,   76.96309284, ...,  -90.60068803,
        -145.75577561,   58.83560098],
       ...,
       [ 151.80809005,   75.19231419,  -90.60068803, ...,  111.50311374,
         171.72232394,  -69.63585277],
       [ 239.13639816,  121.77024102, -145.75577561, ...,  171.72232394,
         276.66488726, -111.54777401],
       [ -96.69853061,  -49.10982093,   58.83560098, ...,  -69.63585277,
        -111.54777401,   45.09929323]])

>>> v = mda.Universe(PSF, DCD, in_memory=True)
>>> a = align.AlignTraj(v, v, select='backbone')
>>> p2 = pca.PCA(v, select='backbone', align=False).run()
>>> p2.mean
array([14.35063446,  7.27347285, -8.7152488 , ..., 10.34658551,
       16.53386494, -6.67638795])
>>> p2.cov
array([[ 1.0575943 , -0.45679874,  0.32512697, ...,  1.79730212,
        -0.58115197,  0.09960816],
       [-0.45679874,  0.32433888, -0.0863796 , ..., -0.83912558,
         0.27184243, -0.04867028],
       [ 0.32512697, -0.0863796 ,  0.22448411, ...,  0.50199753,
        -0.17349537,  0.04935935],
       ...,
       [ 1.79730212, -0.83912558,  0.50199753, ...,  3.3476562 ,
        -1.11032184,  0.15410798],
       [-0.58115197,  0.27184243, -0.17349537, ..., -1.11032184,
         0.47796246, -0.02327182],
       [ 0.09960816, -0.04867028,  0.04935935, ...,  0.15410798,
        -0.02327182,  0.06560988]])
```

**Currently version of MDAnalysis**

- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`) 0.20.2-dev
- Which version of Python (`python -V`)? 3.7
- Which operating system? MacOS

**Tests**

Tests to check that behaviour is fixed should:
* check that `self.mean` is not 0 when `align=True`
* check that the RMSD has lowered after alignment
  * this is probably easiest done by comparing the covariance matrix `self.cov` of PCA done on an unaligned trajectory with `align=True`, and PCA done on a pre-aligned trajectory

### PR DESCRIPTION
**Pull request by hmacdope**
Fixes #2510

Changes made in this Pull Request:
 -PCA now always calculated mean positions, according to recommendations of @lilyminium.
- Tests to come

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [X] Issue raised/referenced?

### CODE REVIEW

**Comment by hmacdope** [comment]
Tests added in covering:
- shape of the mean array
- correct computation of mean positions
- correct alignment using align=True

File: testsuite/MDAnalysisTests/analysis/test_pca.py [hunk 1]
```diff
@@ -126,3 +127,23 @@ def test_cosine_content():
...
+    v = mda.Universe(PSF, DCD, in_memory=True)
+    a = align.AlignTraj(v, v, select=SELECTION).run()
+    coords = v.trajectory.timeseries(v.select_atoms(SELECTION), order='fac')
+    p = PCA(u, select=SELECTION, align=True).run()
+    assert_almost_equal(p.mean, coords.mean(axis=0).ravel(), decimal=5)
+
+def test_alignment(u):
+    v = mda.Universe(PSF, DCD, in_memory=True)
+    a = align.AlignTraj(v, v, select=SELECTION).run()
+    p1 = PCA(u, select=SELECTION, align=True).run()
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_pca.py]: Can use `PCA_aligned` here instead.

**Comment by orbeckst** [comment]
@VOD555 does this PR solve the problem that you had with PCA #2560 ?

File: testsuite/MDAnalysisTests/analysis/test_pca.py [hunk 2]
```diff
@@ -126,3 +127,23 @@ def test_cosine_content():
     dot = pca_random.transform(rand.atoms)
     content = cosine_content(dot, 0)
     assert_almost_equal(content, .99, 1)
+
+def test_mean_shape(u):
+    p = PCA(u, select=SELECTION, align=True).run()
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply testsuite/MDAnalysisTests/analysis/test_pca.py]: Fixed :)

File: testsuite/MDAnalysisTests/analysis/test_pca.py [hunk 3]
```diff
@@ -126,3 +127,23 @@ def test_cosine_content():
...
+    atoms = u.select_atoms(SELECTION)
+    assert_equal(p.mean.shape[0], atoms.n_atoms * 3)
+
+def test_mean(u):
+    v = mda.Universe(PSF, DCD, in_memory=True)
+    a = align.AlignTraj(v, v, select=SELECTION).run()
+    coords = v.trajectory.timeseries(v.select_atoms(SELECTION), order='fac')
+    p = PCA(u, select=SELECTION, align=True).run()
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply testsuite/MDAnalysisTests/analysis/test_pca.py]: Fixed 👍

File: testsuite/MDAnalysisTests/analysis/test_pca.py (hunk 1 above)
Comment (hmacdope) [reply testsuite/MDAnalysisTests/analysis/test_pca.py]: fixed 👍

File: testsuite/MDAnalysisTests/analysis/test_pca.py [hunk 4]
```diff
@@ -126,3 +132,24 @@ def test_cosine_content():
...
+    a = align.AlignTraj(v, v, select=SELECTION).run()
+    coords = v.trajectory.timeseries(v.select_atoms(SELECTION), order='fac')
+    assert_almost_equal(pca_aligned.mean, coords.mean(axis=0).ravel(), decimal=5)
+
+def test_alignment(pca_aligned, u):
+    v = mda.Universe(PSF, DCD, in_memory=True)
+    a = align.AlignTraj(v, v, select=SELECTION).run()
+    pca_pre_align = PCA(v, select=SELECTION, align=False).run()
+    assert_almost_equal(pca_aligned.mean, pca_pre_align.mean)
+    assert_almost_equal(pca_aligned.cov, pca_pre_align.cov)
```
Comment (hmacdope) [inline testsuite/MDAnalysisTests/analysis/test_pca.py]: Thanks for the feedback @orbeckst, much appreciated.
I think this is the kind of test @lilyminium requested, ie
- comparison of covariance matrix of align=True PCA to align=False PCA on prealigned trajectory.
However I may have misinterpreted, please let me know if I have.

File: testsuite/MDAnalysisTests/analysis/test_pca.py [hunk 5]
```diff
@@ -46,6 +47,11 @@ def pca(u):
     u.transfer_to_memory()
     return PCA(u, select=SELECTION).run()
 
+@pytest.fixture(scope='module')
+def pca_aligned(u):
+    u.transfer_to_memory()
```
Comment (lilyminium) [inline testsuite/MDAnalysisTests/analysis/test_pca.py]: Is this necessary? I think PCA fits per frame when aligning.

File: testsuite/MDAnalysisTests/analysis/test_pca.py [hunk 6]
```diff
@@ -24,10 +24,11 @@
 
 import numpy as np
 import MDAnalysis as mda
+from MDAnalysis.analysis import align
 from MDAnalysis.analysis.pca import PCA, cosine_content
 
 from numpy.testing import (assert_almost_equal, assert_equal,
-                           assert_array_almost_equal)
+                           assert_array_almost_equal, assert_raises)
```
Comment (lilyminium) [inline testsuite/MDAnalysisTests/analysis/test_pca.py]: I don’t think you use `assert_raises`?

File: testsuite/MDAnalysisTests/analysis/test_pca.py (hunk 4 above)
Comment (lilyminium) [reply testsuite/MDAnalysisTests/analysis/test_pca.py]: Your interpretation is right, thanks @hmacdope! But I think it’s a stronger test if you don’t transfer `u` into memory before running PCA with `align=True`.

File: package/MDAnalysis/analysis/pca.py [hunk 7]
```diff
@@ -161,7 +161,11 @@ class PCA(AnalysisBase):
     Notes
     -----
     Computation can be speed up by supplying a precalculated mean structure
-
+    
+    .. versionchanged:: 1.0.0
+       Align=True now correctly aligns the trajectory and computes the correct
```
Comment (lilyminium) [inline package/MDAnalysis/analysis/pca.py]: Nitpick: Align should be lowercase.

File: package/CHANGELOG [hunk 8]
```diff
@@ -20,6 +20,8 @@ mm/dd/yy richardjgowers, kain88-de, lilyminium, p-j-smith, bdice, joaomcteixeira
   * 0.21.0
 
 Fixes
+  * pca(align=True) now correctly aligns the trajectory and computes the
```
Comment (lilyminium) [inline package/CHANGELOG]: Nitpick: PCA is all-caps.

File: testsuite/MDAnalysisTests/analysis/test_pca.py (hunk 6 above)
Comment (hmacdope) [reply testsuite/MDAnalysisTests/analysis/test_pca.py]: 👍

File: package/MDAnalysis/analysis/pca.py (hunk 7 above)
Comment (hmacdope) [reply package/MDAnalysis/analysis/pca.py]: 👍

File: package/CHANGELOG (hunk 8 above)
Comment (hmacdope) [reply package/CHANGELOG]: 👍

File: testsuite/MDAnalysisTests/analysis/test_pca.py (hunk 5 above)
Comment (hmacdope) [reply testsuite/MDAnalysisTests/analysis/test_pca.py]: I think it is for speed purposes, however i have removed it as suggested as it strengthens the test.

File: package/CHANGELOG [hunk 9]
```diff
@@ -20,6 +20,8 @@ mm/dd/yy richardjgowers, kain88-de, lilyminium, p-j-smith, bdice, joaomcteixeira
   * 0.21.0
 
 Fixes
+  * PCA(align=True) now correctly aligns the trajectory and computes the
```
Comment (orbeckst) [inline package/CHANGELOG]: add your GitHub handle to the author list for "0.21.0"

Comment (hmacdope) [reply package/CHANGELOG]: 👍

**Comment by orbeckst** [comment]
@hmacdope this looks excellent, we're now just waiting for @lilyminium and Travis.

**Comment by orbeckst** [comment]
(Also, you have a great avatar picture.)

**Comment by hmacdope** [comment]
Thanks, great to finally contribute :)

**Comment by lilyminium** [comment]
Thanks @hmacdope!

**Comment by VOD555** [comment]
@orbeckst This solved the problem. I've tested this with my own trajectory, and it gave me results similar to VMD's. 
Thanks @hmacdope !

**Comment by hmacdope** [comment]
Travis is failing for the 3.6 builds? Appears numpy cannot be found.

**Comment by orbeckst** [comment]
I restarted the failing macOS Python 3.6 Travis job. If this doesn't help we'll have to think. The error was weird because supposedly, numpy 1.17 had been installed.

**Comment by hmacdope** [comment]
Build worked up until yesterday! Very strange indeed. A quick flick through some of the active PRs (eg #2577) show similar failure.

**Pull request merged**